### PR TITLE
Feature/Authentication

### PR DESCRIPTION
**Pull request by treznick**
This PR introduces the `/v1/user_token` endpoint. A successful post with an authentication header matching the following template:
```
{api_key}:{api_secret}
```

will result in a successful response of status 201, with a response body containing a valid JWT with the `sub` claim.

To support this feature, this PR adds a `User` model, which contains the following attributes:
```
email: string, unique
password_digest: string, unique, actually the api_secret
api_key: friendly id for initial authentication to the api
```

Creation of the users happens from a rake task:
```
rake user:create[email]
```

### CODE REVIEW

**Comment by treznick** [comment]
NB: before we can merge this we'll need to turn on HTTPS for all requests.

Also some of these commits can be backported to a cleanup branch prior to merge.

File: app/controllers/v1/user_token_controller.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    private
+
+    def auth_params
+      token = TokenRequest.new(request)
+
+      # NOTE: Knock wants you to get your token using a username and password
+      # from a post body. This is bad from a security standpoint. Also, to avoid
+      # monkey patching a lot of places, we just return a hash that has the
+      # :email and :password keys, even though they're really the :api_key and
+      # :unencrypted_api_secret
```
Comment (treznick) [inline app/controllers/v1/user_token_controller.rb]: edit the line about api_key and unencrypted_api_secret. we call the column password now

**Comment by treznick** [comment]
Also NB this PR does not actually implement any authorization checks into the existing controllers or resources. I suspect we'll want to have some sort of conversation in the upcoming days about just what needs to be protected via authorization. cc @bnd5k @AaronSofaer

File: config/routes.rb [hunk 2]
```diff
@@ -1,6 +1,7 @@
 Rails.application.routes.draw do
 
   namespace :v1, defaults: { format: 'json' } do
+    post 'user_token' => 'user_token#create'
```
Comment (bnd5k) [inline config/routes.rb]: What do you thing about a more generic resource here?  Maybe something not tied directly to the user resource, like `auth/token`, `authentications/token`,  or just have `authorize` map to a post action?

**Comment by bnd5k** [comment]
What happens when a user submits a request with an expired token?  

When we spoke a few weeks back, we talked about an access point that could provide users with an access token, refresh token, and an expiration.  

Knock's access tokens have default lifespan of `1.day`, but I don't see any code in that gem or this PR that addresses how a user would exchange a refresh token for a new access token.  

Is that in the works?

**Comment by bnd5k** [comment]
After this passes code review, can you create a PR for to the documentation app too?

**Comment by bnd5k** [comment]
Thanks for your work on this feature.   

I have some minor complaints about Knock.  Knock seems pretty committed to the idea of passing around an email and password.  I'd prefer to avoid code like ` { email: token.key, password: token.secret }  ` for an approach where we can just call a key a key and a secret a secret.   In the long run, I'd like to remove that dependency.  But for now, I think it'll do just fine.

My Knock quibbles are all small potatoes.  Your code looks well thought-out and well tested.   Thank you!

File: spec/lib/token_request_spec.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+require 'rails_helper'
```
Comment (treznick) [inline spec/lib/token_request_spec.rb]: good catch. `spec_helper` should be enough

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+  describe "#valid?" do
+    let(:user) { described_class.new(user_attributes) }
+    context "given the presence of an email and password" do
+      it "is valid" do
+        expect(user).to be_valid
+      end
+    end
+
+    context "given the absence of an email or password" do
```
Comment (treznick) [inline spec/models/user_spec.rb]: hmm, yeah I think the better fix here is to separate the testing of email absence from password absence to clarify that it's invalid in either case.

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+      it "is invalid" do
+        expect(user).not_to be_valid
+        expect(user.errors[:email]).to be_present
+        expect(user.errors.details[:email]).to match(a_collection_including(a_hash_including(error: :taken)))
+      end
+    end
+  end
+
+  describe ".from_token_request" do
```
Comment (treznick) [inline spec/models/user_spec.rb]: My convention (inherited from a few jobs ago) is to use `#` for instance methods and `.` for class methods, per http://betterspecs.org/#describe, which roughly follows the ruby documentation standards:
http://ruby-doc.org/documentation-guidelines.html

but only roughly:

>Use :: for describing class methods, # for describing instance methods, and use . for example code.

If we want to do `#` for both instance and class methods, then 👍 . We should probably document that somewhere though.

File: app/services/user_creation_service.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# dead simple logger. Expect to see a fake version that's cleverer in specs
```
Comment (treznick) [inline app/services/user_creation_service.rb]: My thought here was to explain why I'm not just putting `puts` directly in the service, but upon retrospection that might be self-evident from the dependency injection in the initalize

File: app/services/user_creation_service.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      logger.log("WRITE IT DOWN NOW")
+    else
+      logger.log("USER WAS INVALID:")
+      logger.log("ERRORS:")
+      logger.log("#{user.errors.full_messages.join('\n')}")
+    end
+    saved
+  end
+
+  def user
```
Comment (treznick) [inline app/services/user_creation_service.rb]: hmm, I think I had left that in there as a hook for testing, but yeah, in a unit test I could just grab `User.last` and be done with it. I'll make it private.

File: config/routes.rb (hunk 2 above)
Comment (treznick) [inline config/routes.rb]: 👍

**Comment by treznick** [comment]
Happy to add documentation :)

I too don't love how Knock sends around authentication params, but short of rolling our own JWT authentication system, I don't see how we can get around knock's opinions on that :/

From what I can tell, the folks behind knock are fairly hesitant to implement refresh tokens, and instead suggest something like sliding sessions to get around token expiration in the middle of, say, a big, paginated, query. 

https://github.com/nsarno/knock/issues/65

I don't _love_ the idea of sliding sessions, since you're basically storing session data on the server side, which kind of runs afoul of REST. 

My gut says that right now it's not immediately clear what the session requirements of our users are, and whether something like sliding sessions would be a good fix, or whether, say, bumping the token expiration time by a day would alleviate most problems. @AaronSofaer Do you have any thoughts on this?

We could definitely pass back the token expiration time in the response, and so I'll gladly fix that up into this PR.

Hope this answers your questions. If necessary, let's have a further discussion in slack about the auth requirements.

**Comment by treznick** [comment]
Actually, looking into how knock does the payload creation, the only point we'd be able to insert the expiration date into the response would be here:
https://github.com/nsarno/knock/blob/master/app/model/knock/auth_token.rb#L28-L30

Which might mean monkey patching the `to_json` method to decrypt its own token to merge in the expires at.

Which, well, while _doable_ starts to sound like some real shenanigans.

@bnd5k what are your thoughts on this? I'm tempted to double down on my earlier assertion that we avoid trying to answer the refresh/session expiration question now, and see in what ways we can answer it when we have a better understanding of the behavior of our api consumers, but I also recognize that we may need this as part of the authentication feature.

**Comment by AaronSofaer** [comment]
The phrase "monkey patching" does not lend itself to the philosophy "build the easiest thing, then iterate".

File: spec/models/user_spec.rb (hunk 5 above)
Comment (bnd5k) [inline spec/models/user_spec.rb]: Interesting.  Thanks for the link.

Let's follow the convention they've outlined here.  I'd rather be able to point future devs to that doc than have to explain our own set of conventions.

**Comment by bnd5k** [comment]
We should not implement sessions, sliding or otherwise.  It makes sense for Knock, because it's geared toward webapps that have an API component.  We've got a straight up and down API.

Let's do this.

- Don't  bother adding expiration to the payload

- Add a test that shows what happens when the client sends a request with an expired token.  This behavior shouldn't be a black hole.   Also, be sure this makes it into your PR for the documentation app. 

- Let's get PR merged and onto staging


This PR introduces Authentication to the app and that's awesome.  We'll worry about fine tuning it later.

File: app/controllers/v1/call_scripts_controller.rb [hunk 8]
```diff
@@ -1,5 +1,6 @@
 module V1
   class CallScriptsController < ApplicationController
+    before_action :authenticate_user, except: [:show, :index]
```
Comment (bnd5k) [inline app/controllers/v1/call_scripts_controller.rb]: We should probably require authenticated for `delete` actions too.

Comment (treznick) [inline app/controllers/v1/call_scripts_controller.rb]: just rebased to `except: [:show, :index]`

**Comment by treznick** [comment]
@bnd5k @AaronSofaer in order to feature spec the JWT side of authentication, and its failures, I had to protect something. Rather than piecemeal the authentication requirements on a per model/per request type basis, I decided that all of our resources should be protected by authentication except for `show` and `index`, meaning that an unauthenticated user can read but cannot create, update or delete.

This may be somewhat aggressive, or, it may also be not aggressive enough. Consider 48a192a to be a starting point for this discussion. If we want to get authentication out the door sooner, then I'm happy to extract that part of the PR into a new PR, so we can start the discussion about what to protect there.

One thing we might also consider is whether we want to store any details about _who_ created a resource, which in turn could be used to scope update permissions.

**Comment by treznick** [comment]
Docs PR coming soon

File: app/controllers/v1/call_scripts_controller.rb (hunk 8 above)
Comment (bnd5k) [inline app/controllers/v1/call_scripts_controller.rb]: 👍

**Comment by bnd5k** [comment]
Thanks for bringing this up: yes, we should be storing who created a given CTA record.  That was one of the original A.C.

We should definitely store the creator for the CTA.  I don't think it's necessary for the contact, location, or call script resource.  @AaronSofaer Would you agree?

**Comment by AaronSofaer** [comment]
> meaning that an unauthenticated user can read but cannot create, update or delete.

That seems totally reasonable. We may want to require users to authenticate before they can read, but we absolutely will not want to allow users to modify any sort of state on the server without authentication.


> We should definitely store the creator for the CTA. I don't think it's necessary for the contact, location, or call script resource. @AaronSofaer Would you agree?

Not sure, but I think they actually are necessary if we're gating modification permissions based on creator. Modifying a call script, functionally speaking, does modify a CTA, because of its association. Same with any other resource. In fact, it modifies a whole bunch (potentially) of CTAs.

The other option that comes to mind is that we could simply not permit modification or deletion of resources, and instead create new resources and allow CTA authors to associate CTAs with these new resources.

**Comment by bnd5k** [comment]
For now, let's store the the CTA's creator.  It's give us meaningful insights into who's using the API.  

We'll save the conversation about of whether someone can modify or delete resources for when we tackle authorization.  The goal for this PR is to introduce Authentication.

**Pull request merged**